LPS-103814

julien opened the pull request:
/cc @diegonvs @p2kmgcl

jpince committed:
LPS-91429 Apply

NorbertKocsis committed:
LPS-98301 Add button to the UI

NorbertKocsis committed:
LPS-98301 Add language key

NorbertKocsis committed:
LPS-98301 Add cleanUpOrphanedPortletPreferences

NorbertKocsis committed:
LPS-98301 Rename action to be more descriptive

NorbertKocsis committed:
LPS-98301 buildLang

drewbrokke committed:
LPS-98301 server-admin-web - lang - makes the portlet preferences lang keys more consistent with each other

drewbrokke committed:
LPS-98301 server-admin-web - (EditServerMVCActionCommand.java) SF - uses Objects.equals

drewbrokke committed:
LPS-98301 server-admin-web - matches param and method names, and sorts the if-else statements

drewbrokke committed:
LPS-98301 server-admin-web - fixes typo in parameter name

jorgediaz-lr committed:
LPS-104116 Check jar version information of lib/versions-ext.xml jar matches with those in lib/*/dependencies.properties files

mateomustapic committed:
LPS-102556 Update dependencies

mateomustapic committed:
LPS-102556 Update AdaptiveMediaProgress component to React

shuyangzhou committed:
LPS-104116 Fix assert, use Assert.assertEquals(), rather than Assert.assertTrue().

shuyangzhou committed:
LPS-104116 SF

jorgediaz-lr committed:
LPS-104116 Normalize version from lib/*/dependencies.properties files

We are using "(\d+(?:\.\d+)+)?(\W)?(.*)" regex in order to obtain numeric version, separator and suffix.

Commit:
LPS-100616 Remove broken icon link, change href

shuyangzhou committed:
LPS-104116 The non-captured group isn't necessary.

shuyangzhou committed:
LPS-104116 No version should be considered as a failure.

shuyangzhou committed:
LPS-104116 Not matching should be considered a faiure.

Commit:
LPS-97294 Add callback to update field

ivicac committed:
COMMERCE-1067 - batch-engine-test - Add BatchEngineExportTaskExecutorTest

ivicac committed:
COMMERCE-1067 - headless-batch-engine-impl - Add logic to ExportTaskResourceImpl

jbalsas committed:
LPS-102556 Simplifies code and keeps it closer to the original to minimize initial differences

Preston-Crary committed:
LPS-104548 Lock down the CTCollection when it's published and maintain correct status / CTPreferences for all users.

Preston-Crary committed:
LPS-104548 Do not delete preferences with the CTCollection, reset them instead.

Commit:
LPS-100997 Make asset change tracked

Commit:
LPS-100997 Fix usages

Commit:
LPS-104414 nullcheck on originalField

Commit:
LPS-104695 Permissions modal

Commit:
LPS-103814 Implement Drag and Drop from Sidebar to PageEditor

Preston-Crary committed:
LPS-104548 Add ability to undo a published collection.

Preston-Crary committed:
LPS-104548 Build Service.

Preston-Crary committed:
LPS-104548 Create copy callback method for CTEventListener.

Preston-Crary committed:
LPS-104548 We publish and copy with SQL. Now delete with SQL to remove one of the ThreadLocals.

pavel-savinov committed:
LPS-104545 Add benchmark property to set number of content pages to create

Preston-Crary committed:
LPS-104548 Update test.

Preston-Crary committed:
LPS-104548 Use correct counter name.

pavel-savinov committed:
LPS-104545 Create specific number of content pages with web content fragment

Preston-Crary committed:
LPS-104548 Add test coverage for undo and undo publication.

Preston-Crary committed:
LPS-104548 Prevent double publishing.

tinatian committed:
LPS-104545 Update unit test

jbalsas committed:
LPS-100134 Exports AssetVocabularyCategoriesSelector component

jbalsas committed:
LPS-100134 Refactors asset-list-web/AutoField component to React

jbalsas committed:
LPS-100134 Refactors asset-publisher-web/AutoField component to React

Commit:
LPS-104414 make ddm-field getValue more robust;

pavel-savinov committed:
LPS-104545 Include fragment script after journal article since we're using default article for our fragment configuration

pavel-savinov committed:
LPS-104545 Keep only web content portlet in the fragment

Commit:
LPS-99441 | LPS-103592 Apply Clay react components on workflow metrics

Commit:
LPS-99441 | LPS-103592 Create PerformanceByAssigneePage component

Commit:
LPS-99441 | LPS-103592 Apply PerformanceByAssigneePage component and improve pages path

Commit:
LPS-99441 | LPS-103592 Migrate ListHeadItem to functional components

Preston-Crary committed:
LPS-104548 Validate all external changes.

Preston-Crary committed:
LPS-104548 Build Service.

Commit:
LPS-99441 | LPS-103592 Create tests to PerformanceByAssigneePage

Preston-Crary committed:
LPS-104548 Fix compile failure in CTEntryBaseImpl.

Preston-Crary committed:
LPS-104548 Build Service.

Preston-Crary committed:
LPS-104548 Allow updates when publishing. Keep the CTEntry modified dates same when publishing. Only update the row when mvccVersions are different. Update tests.

jbalsas committed:
LPS-100134 Removes no longer used legacy vesions of AssetTags and AssetCategories Selectors

ivicac committed:
COMMERCE-1067 - batch-engine-service - Revert back, otherwise it fails with: org.hibernate.PropertyValueException: not-null property references a null or transient value: com.liferay.batch.engine.model.impl.BatchEngineExportTaskImpl.contentBlobModel

Commit:
LPS-99441 | LPS-103592 Apply link component on PerformanceByAssigneeCard

Commit:
LPS-99441 | LPS-103592 Apply new paths in cards link button

Commit:
LPS-99441 | LPS-103592 Regen

Commit:
LPS-99441 | LPS-103592 Add processId on WorkloadByStepCard test case

jonathanmccann committed:
LPS-104643 Make the error message more generic since this exception is thrown during create and update

Commit:
LPS-100997 UpgradeCTModel should take multiple table names

alee8888 committed:
LPS-103488 Create action dropdown to Impersonate User for the Account Users Portlet

Commit:
LRQA-53697 Adding openFormView macro and gotoBack at addFormViewAndValidateFieldLabel macro

Commit:
LRQA-53697 Adding AddMultipleFormViewsToOneObject test case

Commit:
LRQA-53697 SF Adding AddMultipleFormViewsToOneObject test case

alee8888 committed:
LPS-103488 Create action dropdown to Deactivate, Activate, and Delete Account Users

pavel-savinov committed:
LPS-104274 Save created editor soy context to the class field since it is being called more than once

alee8888 committed:
LPS-103488 Add MVC Action Command to support those commands for Account Users

pavel-savinov committed:
LPS-104274 Don't process CSS and HTML when they are empty

pavel-savinov committed:
LPS-104274 Don't process fragment entry link configuration when it is empty

pavel-savinov committed:
LPS-104274 Avoid using streams in the portlet paths processing since normal iteration through the list is faster

pavel-savinov committed:
LPS-104274 Avoid all stream manipulations for the common case when there is only one segment experience id

jbalsas committed:
LPS-100134 Updates yarn.lock

jbalsas committed:
LPS-100134 Removes no longer used EditCategories component

tinatian committed:
LPS-104545 Avoid generation of unused data

tinatian committed:
LPS-104545 Remove unused files

tinatian committed:
LPS-104545 SF, simplify

pavel-savinov committed:
LPS-104274 Don't validate fragment entry link configuration when it is empty

tinatian committed:
LPS-104545 SF, sort

pavel-savinov committed:
LPS-104274 Create document only once for the processing and validation executions

tinatian committed:
LPS-104545 Rename folder

pavel-savinov committed:
LPS-104274 Don't use streams in a common case when there is only one segment experience id

tinatian committed:
LPS-104545 Move all non-ftl file into data folder

pavel-savinov committed:
LPS-104274 Fetch layout page template entry only in the edit mode since it shouldn't be possible to render layout page template directly

pavel-savinov committed:
LPS-104274 Fetch loop-independent variables only once

petershin committed:
LPS-100515 The task name must not contain a colon or forward slash

petershin committed:
LPS-100515 Check .m2-tmp repository

alee8888 committed:
LPS-103488 Add action dropdown when selecting multiple Account Users

alee8888 committed:
LPS-103488 Add functions to handle the action dropdown

alee8888 committed:
LPS-103488 Pass back in the list of Account User Ids to the request

ambrinchaudhary committed:
LPS-98551 Export class to reuse it in other modules

ambrinchaudhary committed:
LPS-98551 Use the metal component and remove AUI use

ambrinchaudhary committed:
LPS-98551 Extract method to reuse it

ambrinchaudhary committed:
LPS-98551 Open the React Preview when an image is uploaded

ivicac committed:
COMMERCE-1067 - batch-engine-api - Rename BatchEngineTaskField to BatchEngineTaskFieldId, to have consuistent names as we already have NestedFieldId that has similar role

ivicac committed:
COMMERCE-1067 - batch-engine-service - Apply BatchEngineTaskField  rename

ivicac committed:
COMMERCE-1067 - batch-engine-test - Apply BatchEngineTaskField  rename

Commit:
LPS-100997 Regen build-service-portlet-asset

Commit:
LPS-100997 Regen build-wsdd-portlet-asset

Preston-Crary committed:
LPS-104674 Extract insert logic into CTRowUtil.

Preston-Crary committed:
LPS-104674 Fix Postgres blob copy reference and delete value problem.

Preston-Crary committed:
LPS-100997 Reuse DatabaseMetaData and DBInspector. Use LoggingTimer with table name.

Preston-Crary committed:
LPS-100997 Combine usages and consider this upgrade as a minor change since we are adding a column.

ling-alan-huang committed:
LPS-84119 Add PropertiesStylingCheck

Commit:
LRQA-52399 Add it for skip portal instance

ling-alan-huang committed:
LPS-84119 Add testcases

ling-alan-huang committed:
LPS-84119 Auto SF

Sharryxue committed:
LRQA-53547 Add macro to edit contact

Sharryxue committed:
LRQA-53547 Update the macro to add Contacts Center on page

pavel-savinov committed:
LPS-104545 Rename to content layout since it is pure backend stuff

Commit:
LRQA-53734 Blacklist elasticsearch 6 related module

ambrinchaudhary committed:
LPS-98551 Update image when it is uploaded to the server

ambrinchaudhary committed:
LPS-98551 Remove AUI dependency

Commit:
LRQA-53697 Updating test case description

dacousalr committed:
LPS-104696 Extend condition to check 3-letter language codes

ambrinchaudhary committed:
LPS-98551 Added missing key

Commit:
LPS-100134 Add missing dependency

Commit:
LPS-100134 SF

Commit:
LPS-100134 Add missing language keys

Commit:
LPS-100134 SF

dacousalr committed:
LPS-104696 Bump the version

matuzalemsteles commented on modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/actions/addFragment.js:
```diff
@@ -0,0 +1,100 @@
+
+				dispatch(addFragmentEntryLinkAndItem({
+					fragmentEntryLink,
+					itemId: `thing-${Date.now()}`,
```
This `itemId` is generated on the front end or it would be generated by the backend now that we are sending the information to the backend.

julien replied:
For now it's generated on the frontend yes, I could have added a package for generating unique IDs (like `uuid` for example) but I thought it wasn't worth it just for a simple test.

As you'll have noticed in order to add an item, you first need to  

1 - Create a fragment entry link (generated on the backend)
2 - Add the item you want to create (update state and re-render)
3 - Ideally send the updated `layoutData` to the server  (backend - not done yet)

Concerning the backend/frontend communication I don't have all the details of the various workflows but maybe @p2kmgcl can help clarify all that.

matuzalemsteles replied:
Thanks @julien!

matuzalemsteles commented on modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/PageEditor.js:
```diff
@@ -95,25 +96,15 @@ function Container({children, item}) {
+	// eslint-disable-next-line
+	const [{canDrop, isDragging}, drop] = useDrop({
+		accept: [LAYOUT_DATA_ITEM_TYPES.column]
+	});
```
will we be able to drop the column too?

julien replied:
Yes, you can drop a column into a row.

We're going to work on this in another task, but there are rules that define what can be dropped where.

matuzalemsteles replied:
Ok, thanks @julien

drewbrokke committed:
LPS-98301 server-admin-web - caches previous value for active from CacheRegistryUtil

matuzalemsteles commented on modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/PageEditor.js:
```diff
@@ -133,3 +125,83 @@ function Row({children, item}) {
+
+	useEffect(() => {
+		if (isMounted()) {
+			// TODO: send updated layoutData to backend
```
I'm a little curious because we're going to update the data here, I am assuming that we will not always update data when we `dispatch` to `store`.

julien replied:
That's the part I also am not sure about.

Basically the intent was to send the updated `layoutData` object each time it changes, to keep it in sync with the backend. This `layoutData` object is fetched but the store and used to render  the various "LayoutData" items (container, column, row, fragment) 

I'm not sure where to do that yet, but we can talk about it.

matuzalemsteles replied:
Hmm, I see I think this will depend on how the server needs to know the information, if we need to send only the fragments that have been moved or the whole `layoutData`... I think depending on the case decide if we will do it here or whenever we update a change in the store, for example move, edit etc...

matuzalemsteles commented on modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/PageEditor.js:
```diff
@@ -133,3 +125,83 @@ function Row({children, item}) {
+	[LAYOUT_DATA_ITEM_TYPES.row]: Row
+};
+
+const LayoutDataItem = React.forwardRef(({item, layoutData}, ref) => {
```
Maybe we won't even need a `forwardRef`. Would that then be used to add the correct `drag`?

julien replied:
Another good question.

Once the items have been "dropped" from the "sidebar" to the "editor" we also need to be able to drag and drop them inside the "editor", so I assumed we would need this `forwardRef` because we need to be able to pass some kind of `ref` (the second element returned by the `useDrag` hook) to the `LayoutDataItem`  component.

matuzalemsteles replied:
Yeah, I thought I'd do everything here but it looks like this is the best place, I was assuming we could only move one Fragment and one Container so I was thinking of adding `useDrag` to their respective components. I think there might be problems with components that we don't want to move in the case of columns.

matuzalemsteles commented on modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/actions/addFragment.js:
```diff
@@ -0,0 +1,100 @@
+			method: 'POST'
+		})
+			.then(response => response.json())
+			.then(fragmentEntryLink => {
```
hey @julien I've been thinking a bit about this if we could do the application offline first, I know there are a lot of things that depend on the backend, but it could help to improve the experience. Offline first here does not mean that you would not need the backend but if the internet goes down the person still keeps editing the fragments and adding them.

For example, adding a `Button` fragment, if this fragment information already exists we can add it to the store without waiting for the backend.

julien replied:
@matuzalemsteles as I've said before the communication with backend is not 100% clear, I guess we could take some time to analyse the previous version of the app but so far I've not been able to add an item without the backend - although I agree it would be nice.

@p2kmgcl what do you think?

matuzalemsteles replied:
@julien, yeah i was thinking better about it, i think we have a big backend dependency to solve the stuff so it's probably not possible to do anything in case we add the sidebar fragment to the page. We can think about it later when we have with all the implemented behaviors maybe.

matuzalemsteles reviewed (commented):
@julien I added some notes!

boton committed:
LPS-104368 Fire a event when image is updated via item selector and apply in preview

boton committed:
LPS-104368 Move events into a separate file

petershin committed:
LPS-100515 Use latest gradle-plugins-app-docker

Commit:
LPS-103365 | LPS-104644 Apply process step filter on "viewAllAssignees" link

ealonso committed:
LPS-100134 buildLang

shuyangzhou committed:
LPS-100997 Array length can never be negative.

pei-jung committed:
LPS-103488 Simplify; no need to declare vars

Preston-Crary committed:
LPS-104523 Walk the tree in a single thread to prevent concurrent updates to the same KaleoInstance row causing StaleObjectStateExceptions.

pei-jung committed:
LPS-103488 Not needed

pei-jung committed:
LPS-103488 Rename

pei-jung committed:
LPS-103488 Refactor

pei-jung committed:
LPS-103488 Sort

pei-jung committed:
LPS-103488 Show proper messages

pei-jung committed:
LPS-103488 Build lang

brianchandotcom committed:
LPS-104674 wordsmith

brianchandotcom committed:
LPS-104545 forgot this one

brianchandotcom committed:
LPS-98301 sort

brianchandotcom committed:
LPS-98301 SF

liferay-continuous-integration committed:
subrepo:ignore Update 'modules/apps/analytics/.gitrepo'.

JoyceWang08211 committed:
LRQA-50561 Add macro for adding web content entry with priority

JoyceWang08211 committed:
LRQA-50561 Add condition for viewing asset priority under WCD

JoyceWang08211 committed:
LRQA-50561 Add case to assert asset priority not reset after publication

JoyceWang08211 committed:
LRQA-50561 Update case name to correct spelling

jorgediaz-lr committed:
LPS-104116 Ignore "RELEASE" suffix

jorgediaz-lr committed:
LPS-104116 Normalize only the suffixes that are specified in the list (FCS, GA, and RC)

We are using "(\d+(\.\d+)+)?(\W)?(RELEASE)?(.*)" regex in order to obtain numeric version, separator and suffix. "RELEASE" suffix is ignored

Following versions are normalized:
 - 1.5rc3 => 1.5 RC3
 - 3.18.2-GA => 3.18.2 GA
 - 1.0-FCS => 1.0 FCS
 - 1.0-rc3 => 1.0 RC3
 - 4.3.22.RELEASE => 4.3.22

Following versions are fine, so they are not modified:
 - 1.5.7
 - 1.1.4c
 - 1.7R4
 - 2.3.0.LIFERAY-PATCHED-1
 - 1.0.5.Final
 - 20070801
 - 02856255a6cc7f

brianchandotcom committed:
LPS-100515 prep next

brianchandotcom committed:
LPS-100515 prep next

brianchandotcom committed:
LPS-104435 prep next

brianchandotcom committed:
LPS-104435 prep next

brianchandotcom committed:
LPS-104435 prep next

Sharryxue committed:
LRQA-53547 Update the macro to view contact from list

Sharryxue committed:
LRQA-53547 Add macro to view contact center container

jorgediaz-lr committed:
LPS-104116 Fix wrong jar versions

boton committed:
LPS-102556 SF

dacousalr committed:
LPS-104696 Add modified files

dacousalr committed:
LPS-104696 Update dependencies

jorgediaz-lr committed:
LPS-104116 ant build-lib-versions

brianchandotcom committed:
LPS-99914 add AccountRole to service.xml

brianchandotcom committed:
LPS-99914 regen

brianchandotcom committed:
LPS-99914 SF

brianchandotcom committed:
LPS-99914 auto SF

brianchandotcom committed:
LPS-99914 readd later

brianchandotcom committed:
LPS-99914 should be singular

inacionery committed:
LPS-99441 | LPS-103592 Simplify to remove duplicate code in LoadingState

Commit:
LPS-104704 Adds support for query params

Commit:
LPS-104704 Expose query utilities

Commit:
LPS-104704 Allows className customization

Commit:
LPS-104704 Expose index of each item

Commit:
LPS-104704 Rename path

ambrinchaudhary committed:
LPS-98551 Do not show  Edit and Info icon if they are empty

ambrinchaudhary committed:
LPS-98551 Set shape of the items elements

Commit:
LPS-104704 Implement upper toolbar and pagination

Commit:
LPS-104704 Adds lang key

Commit:
LPS-104704 Fix styling

liferay-translation committed:
LPS-77699 Update Translations

Commit:
LPS-104704 Adds action menu item to go to details view

inacionery committed:
LPS-99441 | LPS-103592 SF

ambrinchaudhary committed:
LPS-98551 Add button is disabled until an image is successfully uploaded to the server

inacionery committed:
LPS-99441 | LPS-103592 Regen

ambrinchaudhary committed:
LPS-98551 SF

rodrigopaulino committed:
LPS-104599 Fixing SF

Commit:
LPS-104704 Fixes styling of Control Menu

Commit:
LPS-104704 Adds Control Menu to details view and edit view

Commit:
LPS-104704 Fixes styling of edit view

Commit:
LPS-104704 Adds lang key

Commit:
LPS-104720 Semantic version increase too high

AliciaGarciaGarcia committed:
LPS-102824 add info message on seo page to alert user of the custom URL behavior.

AliciaGarciaGarcia committed:
LPS-102824 Build language autogenerated

hhuijser committed:
LPS-104606 Enable check

hhuijser committed:
LPS-104606 Add suppressions

petershin committed:
COMMERCE-1067 Remove unneeded forward slash for paths

petershin committed:
COMMERCE-1067 Generated

vicnate5 committed:
LPS-86667 SF

Commit:
LPS-104704 Adds loading state when navigating between entries

boton committed:
LPS-104718 Add button and transform structure

boton committed:
LPS-104718 Add js to clear the OG image

boton committed:
LPS-104718 Update the field openGraphImageURL to readonly instead of disabled

shuyangzhou committed:
LPS-104523 SF

natocesarrego committed:
LPS-104414 SF

barrosdennys committed:
LRQA-53827 Disabling WorkflowWebContent#EditWebContentInReviewProcess until LPS-104725 is fixed

inacionery committed:
LPS-103368 | LPS-104719 Use a more specific search operation

hhuijser committed:
COMMERCE-1067 SF

Commit:
LPS-104704 Fixes mobile styling

shuyangzhou committed:
LPS-104116 Simplify

shuyangzhou committed:
LPS-104116 Keep special suffix handling logic in java code, rather than regex.

shuyangzhou committed:
LPS-104116 Regen

petershin committed:
LPS-103170 Delete sql\sql.profile-dxp.properties too

koor4356 committed:
LRQA-53724 Add condition to check for test run environment

petershin committed:
LPS-104435 Ran diffArtifact task to check changes

petershin committed:
LPS-104435 Ran printPluginChangelog task

petershin committed:
LPS-104435 Ran fixChangelogs task

Commit:
LPS-104569 Properly serialize ddm-options for backwards compatibility with DDM

hhuijser committed:
LPS-103252 Inline

Commit:
LPS-104569 Fix extraneous row added on every form

Commit:
LPS-104569 Check before using since REST services omit empty values

hhuijser committed:
LPS-103252 Move to BaseCheck

hhuijser committed:
LPS-103252 Add logic to detect cases where we can inline code inside *MapBuider

michaelhashimoto committed:
LRCI-802 Fix database upgrade client within Docker slaves

hhuijser committed:
LPS-103252 Add documentation

hhuijser committed:
LPS-103252 Inline

Sharryxue committed:
LRQA-53547 Add smoke test for Contacts Center

brianchandotcom committed:
LPS-103170 sort

hhuijser committed:
LPS-103252 Add logic to detect cases where we can use *MapBuilder instead of anonymous class

hhuijser committed:
LPS-103252 Fix typo

hhuijser committed:
LPS-103252 Turn off checking anonymous classes until cases are fixed

hhuijser committed:
LPS-103252 Add documentation

brianchandotcom committed:
LPS-98551 SF

brianchandotcom committed:
COMMERCE-1067 prep next

brianchandotcom committed:
LPS-104606 prep next

brianchandotcom committed:
LPS-104606 prep next

Commit:
LPS-104569 SF

Sharryxue committed:
LRQA-53869 Update the locator

brianchandotcom committed:
LPS-103252 prep next

brianchandotcom committed:
LPS-103252 prep next

brianchandotcom committed:
COMMERCE-1067 fix compile

brianchandotcom committed:
LPS-100997 type: git grep "s is empty\""

brianchandotcom committed:
LPS-104116 regen

Commit:
LPS-104755 Link should still work if column value is empty

victorg1991 committed:
LPS-104757 Unify spacer values like in separator and other components

victorg1991 committed:
LPS-102574 Unify spacer values like in separator and other components

brianchandotcom committed:
Revert "LPS-102574 Unify spacer values like in separator and other components"

This reverts commit b2e0c094ecd7330ec71a6ae761eb8958bbdbeb34.

Commit:
LPS-98018 Update alloyeditor to v2.11.2

brianchandotcom committed:
LPS-103252 Hugo SF, I had to do this because it caused another SF error

diegonvs commented:
Hey @matuzalemsteles @julien we can't forget to run the source formatter after rebasing and send to Pablo.

diegonvs commented:
LGTM

julien commented:
@matuzalemsteles @diegonvs Ok I'll rebase on top of master and send to @p2kmgcl

julien commented:
Sent to https://github.com/p2kmgcl/liferay-portal/pull/478

(Has merge conflicts, possibly because p2kmgcl's repo is not up to date with upstream but I think he can still review)
